### ISSUE TITLE
[no-unnecessary-type-assertion] false positive for querySelectorAll()

### ISSUE THREAD

**Issue by Zamralik** [typescript-eslint/typescript-eslint#2817]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

tsconfig.json
```JSON
{
	"compilerOptions": {
		"outDir": "../../public/js/compiled",
		"target": "es2020",
		"lib": ["dom", "es2015", "es2017"],
		"charset": "UTF-8",
		"newLine": "LF",
		"sourceMap": false,
		"allowJs": false,
		"checkJs": false,
		"noEmit": false,
		"esModuleInterop": false,
		"strict": true,
		"alwaysStrict": true,
		"removeComments": true,
		"noImplicitAny": true,
		"noImplicitReturns": true,
		"noImplicitThis": true,
		"strictNullChecks": true,
		"strictFunctionTypes": true,
		"allowUnusedLabels": false,
		"allowUnreachableCode": false,
		"noUnusedLocals": true,
		"noUnusedParameters": true
	},
	"include": [
		"./sources/"
	]
}
```

```TS
(document.body.querySelectorAll('input[name="foo"]') as NodeListOf<HTMLInputElement>).forEach(
	(input: HTMLInputElement): void =>
	{
		input.value = "some value";
	}
);
```
**Expected Result**

No error reported as .querySelectorAll() cannot know exactly what the result could be

**Actual Result**

This assertion is unnecessary since it does not change the type of the expression

**Additional Info**

Removing the assertion has typescript compiler raise an error that the type mismatch

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.8.0` |
| `@typescript-eslint/parser`        | `4.8.0` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.13.0` |
| `node`                             | `14.15.1` |

**Comment by bradzacher** [comment]
Duplicate of #528

This is just a KP we can't do anything about.
`querySelectorAll` is typed as a function that is generic in only the return type.
When you use an explicit assertion on these functions, TS "backfills" the generic to match the assertion:
- Default:
![image](https://user-images.githubusercontent.com/7462525/100333397-66fed800-2f87-11eb-9999-52f8d0a1888f.png)
- With generic annotation:
![image](https://user-images.githubusercontent.com/7462525/100333461-77af4e00-2f87-11eb-98b4-eba0af7d7a92.png)
- With assertion:
![image](https://user-images.githubusercontent.com/7462525/100333492-80a01f80-2f87-11eb-89ff-087800227e3b.png)

Because TS physically backfills the generic type within the backing data structure, when we interrogate the type of the expression - it tells us the same type as the assertion. This causes us to false positive.
There's no way around it and no way for us to fix this.

The workaround is to use the generic annotation instead of an assertion.
```ts
document.body.querySelectorAll<HTMLInputElement>('input[name="foo"]').forEach(
	(input: HTMLInputElement): void =>
	{
		input.value = "some value";
	}
);
```
